### ISSUE TITLE
Issues with consul_iface

### ISSUE THREAD

**Issue by AlJaMa** [ansible-collections/ansible-consul#40]
Hello! 

Thank you for the module! One thing I am not understanding is the "consul_iface" variable. 
Several of my machines do not have the same interface name, so I have attempted to set "consul_iface" to "{{ ansible_default_ipv4.interface }}"  

When your module runs, it then fails with the following error.

```
fatal: [10.11.1.46]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "invocation": {
        "module_args": {
            "dest": "/etc/consul.d/client/config.json", 
            "group": "bin", 
            "owner": "consul", 
            "src": "config_client.json.j2"
        }, 
        "module_name": "template"
    }, 
    "msg": "AnsibleUndefinedVariable: 'dict object' has no attribute u'ansible_officelanbr0'"
}
```
It appears that https://github.com/brianshumate/ansible-consul/blob/master/templates/config_client.json.j2#L29 is properly trying to find the var (see below for the output of `ansible -m setup`

```
       "ansible_officelanbr0": {                                                                                                                           
            "active": true,                                                                                                                                 
            "device": "officelanbr0",                                                                                                                       
            "features": {                                                                                                                                   
                "busy_poll": "off [fixed]",                                                                                                                 
                "fcoe_mtu": "off [fixed]",                                                                                                                  
                "generic_receive_offload": "on",                                                                                                            
                "generic_segmentation_offload": "on",                                                                                                       
                "highdma": "on",                                                                                                                            
                "hw_tc_offload": "off [fixed]",                                                                                                             
                "l2_fwd_offload": "off [fixed]",                                                                                                            
                "large_receive_offload": "off [fixed]",                                                                                                     
                "loopback": "off [fixed]",                                                                                                                  
                "netns_local": "on [fixed]",                                                                                                                
                "ntuple_filters": "off [fixed]",                                                                                                            
                "receive_hashing": "off [fixed]",                                                                                                           
                "rx_all": "off [fixed]",                                                                                                                    
                "rx_checksumming": "off [fixed]",                                                                                                           
                "rx_fcs": "off [fixed]",                                                                                                                    
                "rx_vlan_filter": "off [fixed]",                                                                                                            
                "rx_vlan_offload": "off [fixed]",                                                                                                           
                "rx_vlan_stag_filter": "off [fixed]",                                                                                                       
                "rx_vlan_stag_hw_parse": "off [fixed]",                                                                                                     
                "scatter_gather": "on",                                                                                                                     
                "tcp_segmentation_offload": "on",                                                                                                           
                "tx_checksum_fcoe_crc": "off [fixed]",                                                                                                      
                "tx_checksum_ip_generic": "on",                                                                                                             
                "tx_checksum_ipv4": "off [fixed]",                                                                                                          
                "tx_checksum_ipv6": "off [fixed]",                                                                                                          
                "tx_checksum_sctp": "off [fixed]",                                                                                                          
                "tx_checksumming": "on",                                                                                                                    
                "tx_fcoe_segmentation": "off [requested on]",                                                                                               
                "tx_gre_segmentation": "on",                                                                                                                
                "tx_gso_robust": "off [requested on]",                                                                                                      
                "tx_ipip_segmentation": "on",                                                                                                               
                "tx_lockless": "on [fixed]",                                                                                                                
                "tx_nocache_copy": "off",                                                                                                                   
                "tx_scatter_gather": "on",                                                                                                                  
                "tx_scatter_gather_fraglist": "on",                                                                                                         
                "tx_sit_segmentation": "on",                                                                                                                
                "tx_tcp6_segmentation": "on",                                                                                                               
                "tx_tcp_ecn_segmentation": "on",                                                                                                            
                "tx_tcp_segmentation": "on",                                                                                                                
                "tx_udp_tnl_segmentation": "on",                                                                                                            
                "tx_vlan_offload": "on",                                                                                                                    
                "tx_vlan_stag_hw_insert": "on",                                                                                                             
                "udp_fragmentation_offload": "on",                                                                                                          
                "vlan_challenged": "off [fixed]"                                                                                                            
            },                                                                                                                                              
            "id": "8000.10c37b699bd4",                                                                                                                      
            "interfaces": [                                                                                                                                 
                "vethO17AQ9",                                                                                                                               
                "enp10s0",                                                                                                                                  
                "veth42ECAF"                                                                                                                                
            ],                                                                                                                                              
            "ipv4": {                                                                                                                                       
                "address": "10.11.1.46",                                                                                                                    
                "broadcast": "10.11.1.255",                                                                                                                 
                "netmask": "255.255.255.0",                                                                                                                 
                "network": "10.11.1.0"                                                                                                                      
            },                                                                                                                                              
            "ipv6": [                                                                                                                                       
                {                                                                                                                                           
                    "address": "fe80::12c3:7bff:fe69:9bd4",                                                                                                 
                    "prefix": "64",                                                                                                                         
                    "scope": "link"                                                                                                                         
                }                                                                                                                                           
            ],                                                                                                                                              
            "macaddress": "<removed>",                                                                                                              
            "mtu": 1500,                                                                                                                                    
            "promisc": false,                                                                                                                               
            "stp": false,                                                                                                                                   
            "type": "bridge"                                                                                                                                
        }
```

**Comment by brianshumate** [comment]
Hi there, sorry you're having trouble. This is an interesting use case and the resulting error _could_ be an Ansible bug. I'm going to dig into this and try to come up with a solution.

In the worst case, `consul_iface` might need to be moved into inventory and accessed from there.

**Comment by brianshumate** [comment]
One thing about `ansible_default_ipv4.interface` and why it has not been used before, is that it is not always the correct interface to use when the system has multiple interfaces.

**Comment by groggemans** [comment]
I'm also running into issues with the iface ip lookups because I would like to bootstrap clients without running the playbook against all hosts/servers again. But it seems a lot of changes would be needed to fully support that. (I tested a possible quick fix for this with `setup`, `delegate_to` and `delegate_facts`, but it seems broken again https://github.com/ansible/ansible/issues/17582)

A little more flexibility on defining the bind IP would definitely be nice.

**Comment by brianshumate** [comment]
I am going to have it default to looking in the host inventory for maximum flexibility, but it will still be overridable using Ansible's variable precedence hierarchy and the `CONSUL_IFACE` environment variable.

**Comment by groggemans** [comment]
@brianshumate I'm working on a PR with a few improvements and some cleanup.

One of the things I changed is the default value for `consul_iface`. You changed the code to;
```yaml
consul_iface: "{{ lookup('env','CONSUL_IFACE') | default(hostvars.consul_iface, true) }}"
```
But the `hostvars` value will always overrule the default role value, so it doesn't make sense to do this explicitly. I changed it to the proposed `ansible_default_ipv4.interface` as this makes more sense.

```yaml
consul_iface: "{{ lookup('env','CONSUL_IFACE') | default(ansible_default_ipv4.interface, true) }}"
```

This way there's always a valid iface used, and if that's not the good one, it can be changed by defining one in the host/group vars or env variable. This implementation thus extends the current behavior, and does not replace it.
